Sources/NIOCore/ByteBuffer-binaryEncodedLengthPrefix.swift

```diff
@@ -0,0 +1,315 @@
+    /// - Parameters:
+    ///   - as: The type of integer to be read.
+    ///   - buffer: The buffer to read from.
+    /// - Returns: The integer that was read, or nil if it was not possible to read it.
+    func readInteger<IntegerType: FixedWidthInteger>(
+        as: IntegerType.Type,
+        from buffer: inout ByteBuffer
+    ) -> IntegerType?
+
+    /// Write an integer to a buffer. Move the writer index to after the written integer.
+    /// - Parameters:
+    ///    - integer: The type of the integer to write.
+    ///    - buffer: The buffer to write to.
+    /// - Returns: The number of bytes used to write the integer.
+    func writeInteger<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        to buffer: inout ByteBuffer
+    ) -> Int
+
+    /// An estimate of the number of bytes required to write integers using this strategy.
+    /// Callers may use this to reserve bytes before writing the integer.
+    /// If the actual bytes used by the write function is more or less than this, it may be necessary to shuffle bytes.
+    /// Therefore, an accurate prediction here will improve performance.
+    /// This function will be called from ``ByteBuffer/writeLengthPrefixed(strategy:writeData:)``
+    var requiredBytesHint: Int { get }
+
+    /// Write an integer to a buffer. Move the writer index to after the written integer.
+    /// This function will be called when an integer needs to be written, and some capacity has already been reserved for it.
+    /// Implementers should consider using a less efficient encoding, if possible,to fit exactly within the reserved capacity.
+    /// Otherwise, the caller will need to shift bytes to reconcile the difference.
+    /// It is up to the implementer to find the balance between performance and size.
+    /// - Parameters:
+    ///   - integer: The integer to write
+    ///   - reservedCapacity: The capacity already reserved for writing this integer
+    ///   - buffer: The buffer to write into.
+    /// - Returns: The number of bytes used to write the integer.
+    func writeInteger(
+        _ integer: Int,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int
+}
+
+extension NIOBinaryIntegerEncodingStrategy {
+    @inlinable
+    public var requiredBytesHint: Int { 1 }
+
+    @inlinable
+    public func writeInteger<IntegerType: FixedWidthInteger>(
+        _ integer: IntegerType,
+        reservedCapacity: Int,
+        to buffer: inout ByteBuffer
+    ) -> Int {
+        self.writeInteger(integer, to: &buffer)
+    }
+}
+
+extension ByteBuffer {
+    /// Read a binary encoded integer, moving the `readerIndex` appropriately.
+    /// If there are not enough bytes, nil is returned.
+    @inlinable
+    public mutating func readEncodedInteger<Strategy: NIOBinaryIntegerEncodingStrategy, Integer: FixedWidthInteger>(
+        _ strategy: Strategy,
+        as: Integer.Type = Integer.self
+    ) -> Integer? {
```

Comment:
`readEncodedInteger(.quic)` is a bit odd when using type inference (and is different to `writeEncodedInteger(42, strategy: .quic)`).

I think this should be spelled as `readEncodedInteger(as:strategy:)` which works well with or without type inference and mirrors the write API nicely.